### PR TITLE
pcc-ntupler-74X-v1

### PR DESCRIPTION
**Pull request by capalmer85**
This is the updated version of the pixel cluster counting ntupler from Run1.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+using namespace reco;
+
+class TObject;
+class TTree;
+class TH1D;
+class TFile;
+class RectangularPixelTopology;
+class DetId; 
+
+class PCCNTupler : public edm::EDAnalyzer {
```
Comment (slava77) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h]: all new code should use multi-threading-aware modules as a base.
``` edm::one::EDAnalyzer<> ```
is probably the easiest here

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        cout << "Record \"RunInfoRcd" << "\" does not exist " << endl;
+        fFED1 = 0; 
+        fFED2 = (0x1 << 12) ; 
+    } else {
+        edm::ESHandle<RunInfo> runInfoHandle;
+        iSetup.get<RunInfoRcd>().get(runInfoHandle);
+    }
+    // -- Pixel cluster
+    edm::Handle< edmNew::DetSetVector<SiPixelCluster> > hClusterColl;
+    iEvent.getByLabel(fPixelClusterLabel, hClusterColl);
```
Comment (slava77) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: this should have "consumes" registration and use of getByToken
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#Registering_for_data_access

**Comment by slava77** [comment]
@ktf @Dr15Jones 
does the static analyzer check /test directories?
I was expecting some issues to show up in the static analysis report 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8702/4180/llvm-analysis/

**Comment by gartung** [comment]
@slava77, yes the static analyzer will run on the test directory unless you restrict it with scram.

**Comment by capalmer85** [comment]
@slava77 , can you point me to a simple example of edm::one::EDAnalyzer<> usage?  I just tried replacing edm::EDAnalyzer with edm::one::EDAnalyzer<> and that does not work.

**Comment by slava77** [comment]
would ``` edm::one::EDAnalyzer<edm::one::SharedResources> ``` work?
@Dr15Jones what's appropriate for ntuplers with local TFileService?

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  public:
+    explicit PCCNTupler(const edm::ParameterSet& ps);
+    virtual ~PCCNTupler();
+    virtual void beginJob();
```
Comment (Dr15Jones) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h]: I would also highly recommend adding ```override``` in order to guarantee one gets the signatures of the virtual functions correct.

**Comment by Dr15Jones** [comment]
Since the module wants to see Run transitions it would also have to declare ```edm::one::WatchRuns``` (see https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_one_WatchRuns ).

If one were only filling histograms from ```TFileService``` I think it would be safe to do that simultaneously on different threads. However, I don't think the same is true for filling an TNTuple. So I agree this would need ```edm::one::SharedResources``` where I'd use ```"TFileService"``` as the name of the resource.

**Comment by capalmer85** [comment]
Thank you for the pointers.  I'm unfamiliar with these new classes and its
syntax.  An example would be really helpful for me to follow.

**Comment by Dr15Jones** [comment]
You can find all the documentation which includes simple examples at
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes

**Comment by capalmer85** [comment]
Thanks Chris.  I'll get back to this thread once I have time to review and
digest.

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc (hunk 2 above)
Comment (capalmer85) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: Do you mean just lines 236 or anytime I'm potentially accessing a collection from data?  If just for data then the pileup info is ok as it is, right?

Comment (slava77) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: anything that you will (or may) read from the event should be declared in the constructor as consumes; once the consumed products are declared, you can change getByLabel to getByToken

**Comment by slava77** [comment]
@capalmer85 were there any issues to implement the changes?

**Comment by capalmer85** [comment]
No issues.  I haven't had time to implement the comsumes/get by token part.  I hope to do it this afternoon.  Thanks for checking in.

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h (hunk 3 above)
Comment (capalmer85) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h]: I get an error with override on beginJob and endJob so I left those out.  I added it to the rest.  About to commit.

/afs/cern.ch/work/c/capalmer/private/lumi/CMSSW_7_4_0_pre9/src/RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h:52:18: error: 'virtual void PCCNTupler::beginRun(const edm::Run&, const edm::EventSetup&)' marked override, but does not override
     virtual void beginRun(const edm::Run &, const edm::EventSetup &) override;
                  ^
/afs/cern.ch/work/c/capalmer/private/lumi/CMSSW_7_4_0_pre9/src/RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.h:53:18: error: 'virtual void PCCNTupler::endRun(const edm::Run&, const edm::EventSetup&)' marked override, but does not override
     virtual void endRun(edm::Run const&, edm::EventSetup const&) override;

**Comment by capalmer85** [comment]
@slava77 I've completed the update.  Please give it a look and let me know if there is anything else.

**Comment by Dr15Jones** [comment]
```beginRun``` and ```endRun``` will never be called because that version of ```edm::one::EDAnalyzer<>``` does not have that as virtual functions. This is why the ```override``` is so important. In order to tell the framework you want to know about Run transitions you a have to use the base class

```
edm::one::EDAnalyzer<edm::one::SharedResources,edm::one::WatchRuns>
```

which is described in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_one_WatchRuns

**Comment by capalmer85** [comment]
@Dr15Jones  Yes.  Thank you for the reminder.  Actually WatchLuminosityBlocks is better for what I want.  I don't think I need override here because the standard functions seem fine to me.  Does this all seem ok to you now?

**Comment by Dr15Jones** [comment]
My advise is to always add 'override' to all functions you intend to override from a base class. Its a way to have the compiler always help you.

**Comment by capalmer85** [comment]
Ok.  Now I understand your recommendation better and I've added the override.  Thanks.

**Comment by capalmer85** [comment]
I think I've completed all requested updates.  Can you test and confirm that everything is fine now?

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        nTrk = 0;
+        nPixelClusters.clear();
+        nClusters.clear();
+        layers.clear();
+        BXNo.clear();
+        nGoodVtx.clear();
+    }
+
+    if(sampleType=="MC"){
+        edm::Handle<std::vector< PileupSummaryInfo> > PupInfo;
```
Comment (slava77) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: I don't see a "consumes" for this one.

Comment (capalmer85) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: It is only mc. I understood consumes/getByToken would be for data.

Comment (slava77) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: no, this is irrespective of the input source

Comment (capalmer85) [inline RecoLuminosity/LumiProducer/test/analysis/plugins/PCCNTupler.cc]: Ok.  The twiki used the word "data" in a different way than I expected and I misunderstood the meaning.  Fix coming in a minute.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8702	a59247a
changes in the follow up to the code review comments are implemented.
(jenkins compiled this without errors)

**Pull request merged**